File: uclogic_new_init.c
/* 
 * Copyright (C) 2021 Nicholas CHin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
#include <libusb.h>
#include <stdio.h>

int main(int argc, char **argv)
{
	// Initialization	
	libusb_context *ctx = NULL;
	if (libusb_init(&ctx) != 0)
		printf("Error: Init\n");
	
	libusb_device_handle *dev_h;
	struct libusb_config_descriptor *cfg_desc;
	
	// Read vid and pid from command line argument
	unsigned short vid, pid;
	sscanf(argv[1], "%hx:%hx", &vid, &pid);
	printf("VID: %hx, PID: %hx\n", vid, pid);
	
	// Open specified usb device
	dev_h = libusb_open_device_with_vid_pid(ctx, vid, pid);
	if (dev_h == NULL)
		printf("Error: Open\n");

	if (libusb_get_active_config_descriptor(libusb_get_device(dev_h), &cfg_desc) != 0) {
		printf("Error: Read config descriptor\n");
	}

	// Detach kernel driver for each interface	
	for (int i = 0; i < cfg_desc->bNumInterfaces; i++) {
		printf("Kernel detach: %s\n", libusb_strerror(libusb_detach_kernel_driver(dev_h, i)));
		if (libusb_claim_interface(dev_h, i) != 0) {
			printf("Error: Claim interface\n");
		}
	}

	/* Determine the correct endpoint to send the initialization packet and receive the response.
	 * So far it seems like these endpoints are on the only interface with 2 endpoints, one out
	 * and one in. All other interfaces (the ones that the tablet sends reports over in their
	 * uninitialized state) seem to only have a single in endpoint.
	 */
	unsigned char out_addr, in_addr;
	for (int i = 0; i < cfg_desc->bNumInterfaces; i++) {
		if (cfg_desc->interface[i].altsetting[0].bNumEndpoints == 2) {
			unsigned char addr = cfg_desc->interface[i].altsetting[0].endpoint[0].bEndpointAddress;
			// bit 7 indicates direction. 1 is in, 0 is out
			if (addr & 0x80) {
				in_addr = addr;
				out_addr = cfg_desc->interface[i].altsetting[0].endpoint[1].bEndpointAddress;
			} else {
				out_addr = addr;
				in_addr = cfg_desc->interface[i].altsetting[0].endpoint[1].bEndpointAddress;
			}
		}
	}
	
	unsigned char init_packet[] = {0x02, 0xb0, 0, 0, 0, 0, 0, 0};
	unsigned char data_buff[10] = {0};
	for (int code = 0; code <= 6; code++) { 
		int len = 0;
		init_packet[2] = code;
		printf("\nTrying ");
		for (int i = 0; i<8; i++){
			printf("%02X ", init_packet[i]);
		}
		printf("\nSending init packet: %s\n", libusb_strerror(libusb_interrupt_transfer(dev_h, out_addr, init_packet, 8, &len, 0)));
		//printf("%d\n", len);
		printf("Receive packet: %s\n", libusb_strerror(libusb_interrupt_transfer(dev_h, in_addr, data_buff, 10, &len, 3000)));
		//printf("%d\n", len);
		printf("Received data: ");
		for (int i = 0; i<8; i++){
			printf("0x%02X ", data_buff[i]);
		}
		printf("\n");
	}
	// Reattach kernel driver for each interface	
	for (int i = 0; i < cfg_desc->bNumInterfaces; i++) {
		libusb_release_interface(dev_h, 2);
		libusb_attach_kernel_driver(dev_h, 2);
	}
	libusb_free_config_descriptor(cfg_desc);
	libusb_close(dev_h);
	libusb_exit(ctx);
	return 0;
}

File: uclogic_new_init.py
#!/usr/bin/env python3
#
# Copyright (C) 2021 Nicholas Chin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import sys
from os import getuid
import usb.core
import usb.util

key_packets = ['02B002', '02B004']

if getuid() != 0:
    print("This program must be run with root privileges")
    sys.exit()

# Check number of arguments
if len(sys.argv) < 2:
    print("Missing argument.")
    print("Usage: ./uclogic_new_init.py VID:PID")
    sys.exit()

# Help message
if (sys.argv[1] == "-h" or sys.argv[1] == "--help"):
    print("This program requires root privileges.")
    print("Usage: ./uclogic_new_init.py VID:PID")
    print("Please hold the pen on the tablet surface while running")
    sys.exit()

# Parse arguments
try:
    vid, pid = [int(i, 16) for i in sys.argv[1].split(":")]
except ValueError:
    print("Bad argument")
    print("Usage: ./uclogic_new_init.py VID:PID")
    sys.exit()

dev = usb.core.find(idVendor=vid, idProduct=pid)
if dev is None:
    print("Error! Device not found!")
    sys.exit()

cfg = dev.get_active_configuration()
in_endpoints = list()

# Search for OUT endpoint to send the init key to
for i, iface in enumerate(cfg):
    if dev.is_kernel_driver_active(i):
        dev.detach_kernel_driver(i)
    for e in iface.endpoints():
        if e.bEndpointAddress & 0x80 == 0:
            ei = iface[0]
            eo = iface[1]
            iface_num = i
            in_packet_size = ei.wMaxPacketSize
            out_packet_size = eo.wMaxPacketSize
        else:
            in_endpoints.append(e)

# Test all known key packets
for p in key_packets:
    packet = p + '00'*(out_packet_size-len(p)//2)
    print(f"Trying init packet 0x{packet}:")
    eo.write(bytes.fromhex(packet))

    # Read response packet
    data = ei.read(in_packet_size)
    print("Received " + ''.join(format(x, '02X') for x in data))

    # Flush interfaces
    for e in in_endpoints:
        try:
            e.read(in_packet_size, 1000)
        except usb.core.USBTimeoutError:
            continue

    is_initialized = False
    old_deactivated = True

    # Invalid init packets will enable the correct interface, but
    # do not disable the old ones. Check for reports on the old
    # interfaces to determine key validity
    for e in in_endpoints:
        try:
            data = e.read(in_packet_size, 1000)
            if data is not None:
                if e.bEndpointAddress == ei.bEndpointAddress:
                    is_initialized = True
                else:
                    print("Invalid key!\n")
                    old_deactivated = False
                    break
        except usb.core.USBTimeoutError:
            # Skip interfaces which did not send an event
            continue

    if is_initialized and old_deactivated:
        endpoint = "0x{0:02X}".format(eo.bEndpointAddress)
        print(f"Key is 0x{packet}, interface {iface_num}, endpoint {endpoint}")
        break
